File: cpp/graphAlgorithms/ecolor/ecolor_v.cpp
/** @file ecolor_v.cpp
 * 
 *  @author Jon Turner
 *  @date 2015
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

#include "List_d.h"
#include "Graph.h"

namespace grafalgo {

/** Find a minimum edge coloring in a bipartite graph.
 *  The algorithm is based on the alternating path method.
 *  @param g is a reference to the graph
 *  @param color is an array indexed by an edge number; on return
 *  color[e] is the color assigned to edge e
 *  @return the number of colors used
 */
int ecolor_v(Graph& g, int color[]) {
	int Delta = g.maxDegree();

	// avail[u] is a list of available colors at u
	// emap[u][c] is the edge that is colored c at u
	List_d *avail = new List_d[g.n()+1];
	edge* emap[g.n()+1];
	for (vertex u = 1; u <= g.n(); u++) {
		avail[u].resize(Delta);
		emap[u] = new edge[Delta+1];
		for (int c = 1; c <= Delta; c++) {
			avail[u].addLast(c); emap[u][c] = 0;
		}
	}

	// color each edge in turn
	for (edge e = g.first(); e != 0; e = g.next(e)) {
		// first look for a color that is available at both endpoints
		vertex u = g.left(e); vertex v = g.right(e);
		int cu = avail[u].first(); int cv = avail[v].first();
		while (cu != 0 && cv != 0 && cu != cv) {
			if (cu < cv) cu = avail[u].next(cu);
			else	     cv = avail[v].next(cv);
		}
		if (cu != 0 && cv != 0) {
			// cu == cv, is available at both
			color[e] = cu;
			avail[u].remove(cu); avail[v].remove(cv);
			emap[u][cu] = e; emap[v][cv] = e;
			continue;
		}
		// follow alternating (cu,cv) path and flip its colors
		// depends on graph being bipartite
		cu = avail[u].first(); cv = avail[v].first();
		vertex w = v; int c = cu; edge f = e;
		while (emap[w][c] != 0) {
			// f is next edge on path to be colored
			// w is the "leading endpoint of f"
			// c is the color to use for f
			edge ff = emap[w][c];	// next edge in the path
			color[f] = c;
			emap[g.left(f)][c] = f; emap[g.right(f)][c] = f;
			c = (c == cu ? cv : cu);
			w = g.mate(w,ff);
			f = ff;
		}
		// color the last edge and update the avail sets at endpoints
		color[f] = c;
		emap[g.left(f)][c] = f; emap[g.right(f)][c] = f;
		avail[u].remove(cu); avail[v].remove(cv);

		// update available colors at last vertex on path
		avail[w].remove(c);
		c = (c == cu ? cv : cu);
		emap[w][c] = 0;
		int ac = avail[w].first();
		if (c < ac) avail[w].addFirst(c);
		else {
			while (ac != 0 && c > avail[w].next(ac))
				ac = avail[w].next(ac);
			if (ac == 0) avail[w].addLast(c);
			else avail[w].insert(c,ac);
		}
	}
	delete [] avail;
	return Delta;
}

} // ends namespace

File: js/graphAlgorithms/ecolor/ecolorV.mjs
/** @file ecolorV.mjs
 *
 *  @author Jon Turner
 *  @date 2022
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

import { fassert } from '../../common/Errors.mjs';
import List from '../../dataStructures/basic/List.mjs';
import findSplit from '../misc/findSplit.mjs';
import maxflowD from '../maxflow/maxflowD.mjs';

let g;         // shared reference to graph
let avail;     // avail[u] is a List of available colors at u
let emap;      // emap[u][c] is the edge that is colored c at u

let trace;
let traceString;

let recolors;  // number of recoloring operations
let rsteps;    // number of steps in recolor operations
let steps;     // total number of steps

/** Compute a coloring of a bipartite graph using the augmenting
 *  path algorithm from the proof of Vizing's theorem.
 *  @param cg is an undirected bipartite graph
 *  @param traceFlag causes a trace string to be returned when true
 *  @return a pair [ts, stats] where ts is a possibly
 *  empty trace string and stats is a statistics object;
 *  the coloring is returned as integer edge colors in g
 */
export default function ecolorV(cg, traceFlag=false) {
	g = cg; trace = traceFlag;
	let Delta = g.maxDegree();

	traceString = '';
	recolors = rsteps = steps = 0;

	avail = new Array(); avail.push(null);
	emap = new Array(); emap.push(null);
	for (let u = 1; u <= g.n; u++) {
		avail.push(new List(Delta));
		emap.push(new Int32Array(Delta+1));
		for (let c = 1; c <= Delta; c++) {
			avail[u].enq(c); emap[u][c] = 0; steps++;
		}
	}

	// color each edge in turn
	let ecnt = 0;
	for (let e = g.first(); e != 0; e = g.next(e)) {
		let [u,v] = [g.left(e),g.right(e)];
		let c = avail[u].common(avail[v]);
			// returns first color in avail[u] that's also in avail[v];
		if (c) {
			g.color(e,c);	// color is an alias for weight
			if (trace) traceString += `${g.e2s(e,0,1)}:${c} `;
			avail[u].delete(c); avail[v].delete(c);
			emap[u][c] = emap[v][c] = e;
			if (trace && (ecnt++)%12 == 11) traceString += '\n';
			continue;
		}
		if (traceString && ecnt%12 != 0) {
			traceString += '\n'; ecnt = 0;
		}
		recolor(e);
	}

	if (trace) traceString += '\n' + g.toString(1);
	return [traceString, { 'recolors': recolors,
						   'avg path length': ~~(rsteps/recolors),
	 					   'steps': steps }];
}

/** Color an edge by reversing colors on an alternating color path.
 *  @param e is an edge for which no color is available at both endpoints.
 */
function recolor(e) {
	recolors++;
	let [u,v] = [g.left(e),g.right(e)];
	let cu = avail[u].first(); let cv = avail[v].first();
	let f = e; let w = v; let c = cu;
	if (trace) traceString += '[';
	while (emap[w][c] != 0) {
		// f is next edge on path to be colored
		// w is the "leading endpoint" of f
		// c is the color to use for f
		steps++; rsteps++;
		let ff = emap[w][c];	// next edge in the path
		g.color(f,c); emap[g.left(f)][c] = emap[g.right(f)][c] = f;
		if (trace) traceString += `${g.e2s(f,0,1)}:${c} `;
		f = ff; w = g.mate(w,ff); c = (c == cu ? cv : cu);
	}
	// color the last edge and update the avail sets at endpoints
	g.color(f,c); emap[g.left(f)][c] = emap[g.right(f)][c] = f;
	avail[u].delete(cu); avail[v].delete(cv);
	if (trace) traceString += `${g.e2s(f,0,1)}:${c}]\n`;

	// update available colors at last vertex on path
	avail[w].delete(c);
	c = (c == cu ? cv : cu);
	emap[w][c] = 0; avail[w].push(c);
}
